### PR TITLE
Do not forward pages with parameters by default

### LINKED ISSUE #488: Calling unpublished sibling page of forward target causes redirect instead of immediate 404
Tested in Contao 4.4.39

## How to reproduce

* Activate `folderUrl`

* Create a page structure like this:
  ```
  foo
  foo/bar
  foo/baz
  ```

* Make `foo` an internal redirect (page type `forward`) to `foo/bar`.

* Unpublish `foo/baz`.

* Call `example.test/foo/baz.html`

Result: Redirect to `example.test/foo/bar/baz.html` (404).
Expected: Immediate 404.

## Cause

In `Frontend::getPageIdFromUrl()`, `PageModel::findByAliases()` is called[1], which purposefully does not find unpublished pages[2]. As a result, `baz` is used as an `auto_item` value added to `foo` instead. Then, as `foo` is a `forward` page, we get a redirect to `foo/bar` with `baz` added in `PageForward::getForwardUrl()`[3].

Here is some context on where this has come from:
https://github.com/contao/core/issues/4652
The intent is that the alias of an unpublished page does not prevent a fragment of that alias (in this case, `baz`) from being used as an `auto_item` value or full GET variable/value pair. It's nice that we don't have that limitation, but as a result, we have this redirect issue instead. :D

## What now?

I'm not sure how to solve this in a way that doesn't break any other aspect of `folderUrl`. I suppose one could add this to the very end of `PageForward::getForwardUrl()`, just before the `return` statement[4]:
```
$objTargetPage = \PageModel::findByAlias($objPage->alias . $strGet);
if ($objTargetPage !== null)
{
	while ($objTargetPage->next())
	{
		if ($objTargetPage->published == '1')
		{
			continue;
		}

		$objUnpublishedPage = $objTargetPage->current();
		$objUnpublishedPage->loadDetails();
		if ($objUnpublishedPage->language === $objPage->language)
		{
			throw new PageNotFoundException('Page not found');
		}
	}
}
```

It works, but it requires an extra DB request for the target page *plus* `loadDetails()`. Furthermore, far into `PageForward` is an unexpected place to have another check whether we might want to throw a `PageNotFoundException`.

It can also be considered a break in expected behavior, in case somebody actually needs a parameter to be passed via the forward page for some reason (maybe short URLs…?).

Possibly related: https://github.com/contao/contao/issues/316

---

[1] https://github.com/contao/contao/blob/947b6254e4420a22ba58f6baffb125da561dd459/core-bundle/src/Resources/contao/classes/Frontend.php#L136
[2] https://github.com/contao/contao/blob/947b6254e4420a22ba58f6baffb125da561dd459/core-bundle/src/Resources/contao/models/PageModel.php#L509-L513
[3] https://github.com/contao/contao/blob/947b6254e4420a22ba58f6baffb125da561dd459/core-bundle/src/Resources/contao/pages/PageForward.php#L90-L116 
[4] https://github.com/contao/contao/blob/947b6254e4420a22ba58f6baffb125da561dd459/core-bundle/src/Resources/contao/pages/PageForward.php#L123-L125

### PR DESCRIPTION
**Pull request by leofeyer**
Fixes #488

### TODO

- [x] Migration

### CODE REVIEW

File: core-bundle/src/Routing/Page/PageRegistry.php [hunk 1]
```diff
@@ -64,9 +64,13 @@ public function getRoute(PageModel $pageModel): PageRoute
             $path = '';
             $options['compiler_class'] = UnroutablePageRouteCompiler::class;
         } elseif (null === $path) {
-            $path = '/'.($pageModel->alias ?: $pageModel->id).'{!parameters}';
-            $defaults['parameters'] = '';
-            $requirements['parameters'] = $pageModel->requireItem ? '/.+?' : '(/.+?)?';
+            if ('forward' === $pageModel->type && $pageModel->forwardParams) {
```
Comment (fritzmg) [inline core-bundle/src/Routing/Page/PageRegistry.php]:
```suggestion
            if ('forward' === $pageModel->type && !$pageModel->forwardParams) {
```

File: core-bundle/src/Resources/contao/languages/en/tl_page.xlf [hunk 2]
```diff
@@ -74,6 +74,12 @@
       <trans-unit id="tl_page.redirect.1">
         <source>Please choose the redirect type.</source>
       </trans-unit>
+      <trans-unit id="tl_page.forwardParams.0">
+        <source>Forward with parameters</source>
+      </trans-unit>
+      <trans-unit id="tl_page.forwardParams.1">
+        <source>Forward to the target page even if there are query or path parameters.</source>
```
Comment (fritzmg) [inline core-bundle/src/Resources/contao/languages/en/tl_page.xlf]: Not sure if this wording makes it clear, that the redirect will also forward the query or path parameters to the target page. I would prefer

> Forward query and path parameters to the target page.

**Review by fritzmg — COMMENTED** [review]
Tested this with and without legacy routing, including in the presence of a `getPageIdFromUrl` hook and works as intended 👍

File: core-bundle/src/Resources/contao/languages/en/tl_page.xlf (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/languages/en/tl_page.xlf]: But that sounds like

- if the checkbox is checked, requests will be forwarded with parameters
- if the checkbox is not checked, requests will be forwarded without parameters

whereas in the second case, Contao will return a 404 response instead.

Comment (fritzmg) [reply core-bundle/src/Resources/contao/languages/en/tl_page.xlf]: I agree, but the original wording sounds to me as if the query and path parameters are not included in the redirect. Indicating both in the description is probably too technical though.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/languages/en/tl_page.xlf]: >  the original wording sounds to me as if the query and path parameters are not included in the redirect

To me, the label suggests that, not the help text. Shall we change the label to "Always forward"?

Comment (fritzmg) [reply core-bundle/src/Resources/contao/languages/en/tl_page.xlf]: > To me, the label suggests that, not the help text.

I guess that's true. May be leave it then?

Or change both things:

> **Always forward**
>
> Forwards query and path parameters to the target page.

**Review by ausi — DISMISSED** [review]

**Review by fritzmg — DISMISSED** [review]

File: core-bundle/src/Resources/contao/languages/en/tl_page.xlf (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/languages/en/tl_page.xlf]: Changed in 3a5a213f928d6abf8c81ea21bd46d73c89344108.

**Review by ausi — APPROVED** [review]

**Pull request merged**